low-level/framework/messaging/thread_controller.py

```diff
@@ -48,38 +51,158 @@
+        f"{module_name.upper()}>sensor_recovery_interval", recovery_interval)
+    return recovery_count, recovery_interval
 
 
 # Global method used by Thread to capture and log errors.  This must be global.
-def _run_thread_capture_errors(curr_module, sspl_modules, msgQlist,
-                               conf_reader, product):
-    """Run the given thread and log any errors that happen on it.
-    Will stop all sspl_modules if one of them fails."""
-    try:
-        # Each module is passed a reference list to message queues so it can transmit
-        #  internal messages to other modules as desired
-        curr_module.start_thread(conf_reader, msgQlist, product)
-
-    except BaseException as ex:
-        logger.critical(
-            "SSPL-LL encountered a fatal error, terminating service Error: %s" % ex)
-        logger.exception(ex)
-
-        # Populate an actuator response message and transmit back to HAlon
-        error_msg = "SSPL-LL encountered an error, terminating service Error: " + \
-                    ", Exception: " + logger.exception(ex)
-        json_msg = ThreadControllerMsg(curr_module.name(), error_msg).getJson()
-
-        if product.lower() in [x.lower() for x in enabled_products]:
-            self._write_internal_msgQ(EgressProcessor.name(), json_msg)
-        elif product.lower() in [x.lower() for x in cs_legacy_products]:
-            self._write_internal_msgQ(
-                PlaneCntrlRMQegressProcessor.name(), json_msg)
-
-        # Shut it down, error is non-recoverable
-        for name, other_module in list(sspl_modules.items()):
-            if other_module is not curr_module:
-                other_module.shutdown()
+def execute_thread(module, msgQlist, conf_reader, product, resume=True):
+    """
+    Run module as a thread. Recover the module if any error during
+    initialization and run time of the module.
+
+    If recovery count>0,
+        module will be recovered from failure until the maximum recovery
+        attempt. If not recoverable, corresponding module will be shutdown
+        and failure alert will be raised due to its impact.
+    If recovery count=0,
+        no recovery attempt will be made.
+    """
+    module_name = module.name()
+
+    # Initialize persistent cache for sensor status
+    per_data_path = os.path.join(
+        module_cache_dir, f"{module_name.upper()}_{node_id}")
+    if not os.path.isfile(per_data_path):
+        module_persistent_data[module_name] = {}
+        store.put(module_persistent_data[module_name], per_data_path)
+
+    is_sensor_thread = False
+    recovery_count = recovery_interval = 0
+    if isinstance(module, SensorThread):
+        recovery_count, recovery_interval = _get_recovery_config(module_name)
+        is_sensor_thread = True
+
+    attempt = 0
+
+    while attempt <= recovery_count:
+        attempt += 1
+        try:
+            # Suspend module threads
+            if resume == False:
+                module.suspend()
```

Comment:
Check this before 'while' and skip looping altogether

Reply:
Done.